1st draft nft-storage (#1868)
XRPLF/xrpl-dev-portal: content/concepts/tokens/nft-storage.md

---
html: nft-storage.html
parent: non-fungible-tokens.html
blurb: Storage options for the payload of your NFT.
labels:
  - Non-fungible Tokens, NFTs
---
# NFT Payload Storage

> NFTs are created on the block chain. But the payload of the NFT, including media, metadata, and attributes can be stored in a variety of ways, including on-chain, decentralized off-chain, and centralized off-chain.

Reviewer: ```suggestion
NFTs are created on the blockchain. But the payload of the NFT, including media, metadata, and attributes can be stored in a variety of ways, including on-chain, decentralized off-chain, and centralized off-chain.
```

Reviewer: The term "off-chain" doesn't appear to have been used in the XRPL documentation so far; "on/off-ledger" has been used instead.

## On-chain

If your data is smaller than 256 bytes, you can consider using a `data://` URI and embedding it directly in the URI field. This has the advantage of storing the data in a reliable, persistent, and responsive database.

## Decentralized, off XRPL

You can use any of the existing decentralized storage solutions for your NFT metadata.

IPFS and Arweave offer solutions for decentralization. However, fetching the metadata efficiently can be a problem. Querying IPFS or Arweave directly to fetch metadata is not fast enough for modern websites that require immediate responses from users that are scrolling through multiple pages of NFTs that include high-quality media.

> For a list of cloud storage solutions, the blog post [NFT Payload Storage](https://xrpl.org/blog/2023/nft-payload-storage.html).

Reviewer: ```suggestion
See the blog post [NFT Payload Storage](https://xrpl.org/blog/2023/nft-payload-storage.html) for some examples of cloud storage solutions.
```
